### PR TITLE
*: update backoff

### PR DESCRIPTION
**Pull request by disksing**
@qiuyesuifeng

### CODE REVIEW

File: kv.go [hunk 1]
```diff
@@ -167,70 +162,59 @@ func (s *tikvStore) CurrentVersion() (kv.Version, error) {
...
-		// NotLeader error need update region and getRegion again.
-		if resType != nil && resType.GetType() == pb.ResultType_NotLeader {
-			leaderInfo := resType.GetLeaderInfo()
-			if leaderInfo == nil {
-				return nil, errors.New("leader info not found")
-			}
-			s.regionCache.UpdateLeader(leaderInfo.GetRegionId(), leaderInfo.GetLeader().GetId())
+		if notLeader := extractNotLeader(resp); notLeader != nil {
+			log.Warnf("tikv reports `NotLeader`: %s, retry later", notLeader.String())
+			s.regionCache.UpdateLeader(notLeader.GetRegionId(), notLeader.GetLeader().GetId())
```
Comment (qiuyesuifeng) [inline kv.go]: Can we guarantee that the `notLeader.GetLeader()` value is not nil?

File: kv.go [hunk 2]
```diff
@@ -167,70 +162,59 @@ func (s *tikvStore) CurrentVersion() (kv.Version, error) {
...
 			return nil, errors.Trace(err)
 		}
-		log.Debugf("addr: %s %d", region.GetAddress(), region.peer.GetId())
 		client, err := s.getClient(region.GetAddress())
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
 		req.Context = region.GetContext()
-		resp, err = client.SendKVReq(req)
+		resp, err := client.SendKVReq(req)
```
Comment (qiuyesuifeng) [inline kv.go]: It is better to rename err variable, or modify `for loop` condition `err == nil` to make it clear.

File: kv.go (hunk 1 above)
Comment (disksing) [inline kv.go]: It is guaranteed by tikv server, but still we should check it. I'll fix it soon. @qiuyesuifeng

Comment (disksing) [inline kv.go]: @qiuyesuifeng Just reconsidered it.
Thanks to pb's optional values, if `GetLeader()` returns `nil`, it won't panic but return a default value `0`.

**Comment by qiuyesuifeng** [comment]
LGTM

File: backoff.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	FullJitter
+	// Also randomized, but prevents very short sleeps.
+	EqualJitter
+	// Increase the maximum jitter based on the last random value.
+	DecorrJitter
+)
+
+// BackOff Implements exponential backoff with optional jitters.
+// See: http://www.awsarchitectureblog.com/2015/03/backoff.html
+func NewBackoff(retry, base, cap, jitter int) func() error {
```
Comment (siddontang) [inline backoff.go]: should we need all jitters?

Comment (disksing) [inline backoff.go]: @siddontang We don't know yet. Later I'll add backoff for lock wait, pd calls, etc. Seems some of the jitters could be helpful.

Comment (siddontang) [inline backoff.go]: why not move this to ti-common repository? 

@shenli

**Comment by siddontang** [comment]
LGTM

**Pull request merged**